File: app/src/main/java/org/radarcns/detail/PrivacyPolicyFragment.java
package org.radarcns.detail;

import android.app.Fragment;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import org.radarcns.android.RadarConfiguration;
import org.radarcns.android.auth.AppAuthState;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;

import static org.radarcns.android.auth.portal.ManagementPortalClient.BASE_URL_PROPERTY;
import static org.radarcns.android.auth.portal.ManagementPortalClient.PRIVACY_POLICY_URL_PROPERTY;
import static org.radarcns.detail.DetailInfoActivity.PRIVACY_POLICY;
import static org.radarcns.detail.DetailMainActivityView.MAX_USERNAME_LENGTH;
import static org.radarcns.detail.DetailMainActivityView.truncate;

public class PrivacyPolicyFragment extends Fragment implements View.OnClickListener {
    private static final Logger logger = LoggerFactory.getLogger(PrivacyPolicyFragment.class);
    private TextView dataCollectionPolicyLink;
    private OnFragmentInteractionListener mListener;
    private AppAuthState authState;
    private String privacyPolicyUrl;

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment PrivacyPolicyFragment.
     */
    public static PrivacyPolicyFragment newInstance(AppAuthState state) {
        PrivacyPolicyFragment fragment = new PrivacyPolicyFragment();
        Bundle args = new Bundle();
        args = state.addToBundle(args);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null) {
            authState = AppAuthState.Builder.from(args).build();
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_privacy_policy, container, false);

        dataCollectionPolicyLink = view.findViewById(R.id.dataCollectionDescriptionStatement);
        dataCollectionPolicyLink.setOnClickListener(this);

        TextView generalPrivacyPolicyLink = view.findViewById(R.id.generalPrivacyPolicyStatement);
        Serializable privacyPolicy = authState.getProperty(PRIVACY_POLICY_URL_PROPERTY);
        if (privacyPolicy != null) {
            privacyPolicyUrl = privacyPolicy.toString();
            generalPrivacyPolicyLink.setOnClickListener(this);
            generalPrivacyPolicyLink.setVisibility(View.VISIBLE);
            generalPrivacyPolicyLink.setText(getString(R.string.general_privacy_policy));
        } else {
            generalPrivacyPolicyLink.setVisibility(View.INVISIBLE);
        }

        Button acceptedButton = view.findViewById(R.id.accept_privacy_policy_button);
        acceptedButton.setOnClickListener(this);

        String projectId = getString(R.string.study_id_message, authState.getProjectId());
        TextView mProjectId = view.findViewById(R.id.inputProjectId);
        mProjectId.setText(projectId);

        String userId = getString(R.string.user_id_message, truncate(authState.getUserId(), MAX_USERNAME_LENGTH));
        TextView mUserId = view.findViewById(R.id.inputUserId);
        mUserId.setText(userId);

        String baseUrl = (String) authState.getProperties().get(BASE_URL_PROPERTY);
        TextView acceptanceStatement = view.findViewById(R.id.policyAcceptanceStatement);
        acceptanceStatement.setText(getString(R.string.policy_acceptance_message, baseUrl));
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    private synchronized String getDataCollectionUrl() {
        return RadarConfiguration.getInstance().getString(PRIVACY_POLICY);
    }

    private String updatePrivacyStatement() {
        final String url = getDataCollectionUrl();
        logger.info("Setting privacy policy {}", url);

        if (url != null) {
            dataCollectionPolicyLink.setVisibility(View.VISIBLE);
            dataCollectionPolicyLink.setText(R.string.collected_data_description);
        } else {
            dataCollectionPolicyLink.setVisibility(View.INVISIBLE);
        }
        return url;
    }


    public void openUrl(String url) {
        if (url != null) {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
        }
    }

    public void acceptPrivacyPolicy() {
        logger.info("Policy accepted. Redirecting to DetailedMainView...");
        mListener.onAcceptPrivacyPolicy(authState);
    }

    @Override
    public void onClick(View view) {
        int viewId = view.getId();
        switch (viewId) {
            case R.id.accept_privacy_policy_button:
                acceptPrivacyPolicy();
                break;
            case R.id.dataCollectionDescriptionStatement:
                openUrl(updatePrivacyStatement());
                break;
            case R.id.generalPrivacyPolicyStatement:
                openUrl(privacyPolicyUrl);
                break;
        }
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        void onAcceptPrivacyPolicy(AppAuthState state);
    }
}

File: app/src/main/java/org/radarcns/detail/PrivacyPolicyFragment.kt
package org.radarcns.detail

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Html
import android.text.Spanned
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import org.radarbase.android.RadarApplication.Companion.radarConfig
import org.radarbase.android.RadarConfiguration.Companion.PROJECT_ID_KEY
import org.radarbase.android.RadarConfiguration.Companion.USER_ID_KEY
import org.radarbase.android.auth.AppAuthState
import org.radarbase.android.auth.AuthService.Companion.BASE_URL_PROPERTY
import org.radarbase.android.auth.AuthService.Companion.PRIVACY_POLICY_URL_PROPERTY
import org.radarcns.detail.InfoActivity.Companion.PRIVACY_POLICY
import org.radarcns.detail.databinding.FragmentPrivacyPolicyBinding
import org.slf4j.LoggerFactory

class PrivacyPolicyFragment : Fragment() {
    private var mListener: OnFragmentInteractionListener? = null
    private var privacyPolicyUrl: String? = null
    private var projectId: String? = null
    private var userId: String? = null
    private var baseUrl: String? = null
    private var binding: FragmentPrivacyPolicyBinding? = null

    private var dataCollectionUrl: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val args = checkNotNull(arguments) { "Cannot start without AppAuthState arguments" }

        projectId = args.getString(PROJECT_ID_KEY)
        userId = args.getString(USER_ID_KEY)
        baseUrl = args.getString(BASE_URL_PROPERTY)
        privacyPolicyUrl = args.getString(PRIVACY_POLICY_URL_PROPERTY)
        dataCollectionUrl = args.getString(PRIVACY_POLICY)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = FragmentPrivacyPolicyBinding.inflate(inflater, container, false)
        .also { binding = it }
        .root

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding?.apply {
            dataCollectionDescriptionStatement.setOnClickListener { openUrl(updatePrivacyStatement()) }

            generalPrivacyPolicyStatement.apply {
                visibility = if (privacyPolicyUrl != null) {
                    setOnClickListener { openUrl(privacyPolicyUrl) }
                    View.VISIBLE
                } else {
                    View.INVISIBLE
                }
            }

            acceptPrivacyPolicyButton.setOnClickListener { acceptPrivacyPolicy() }
            rejectPrivacyPolicyButton.setOnClickListener { rejectPrivacyPolicy() }

            inputProjectId.text = projectId
            inputUserId.text = userId

            val baseUrl = baseUrl ?: "Unknown server"
            consentPrivacyPolicy.apply {
                text = fromHtml(R.string.consent_privacy_policy, Html.FROM_HTML_MODE_LEGACY)
                setOnCheckedChangeListener { _, isChecked ->
                    acceptPrivacyPolicyButton.isEnabled = consentCollectedData.isChecked && consentServer.isChecked && isChecked
                }
            }
            consentCollectedData.apply{
                text = fromHtml(R.string.consent_collected_data, Html.FROM_HTML_MODE_LEGACY)
                setOnCheckedChangeListener { _, isChecked ->
                    acceptPrivacyPolicyButton.isEnabled = consentPrivacyPolicy.isChecked && consentServer.isChecked && isChecked
                }
            }
            consentServer.apply{
                text = fromHtml(R.string.consent_server, baseUrl, Html.FROM_HTML_MODE_LEGACY)
                setOnCheckedChangeListener { _, isChecked ->
                    acceptPrivacyPolicyButton.isEnabled = consentPrivacyPolicy.isChecked && consentCollectedData.isChecked && isChecked
                }
            }
            inputDestinationUrl.text = baseUrl
        }
    }


    @Suppress("DEPRECATION")
    private fun fromHtml(@StringRes resourceId: Int, vararg parameters: Any): Spanned {
        val s = getString(resourceId, *parameters)
        return Html.fromHtml(s, Html.FROM_HTML_MODE_LEGACY)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
    }

    private fun updatePrivacyStatement(): String? {
        val url = dataCollectionUrl
        logger.info("Setting privacy policy {}", url)

        binding?.dataCollectionDescriptionStatement?.apply {
            if (url != null) {
                visibility = View.VISIBLE
                setText(R.string.collected_data_description)
            } else {
                visibility = View.INVISIBLE
            }
        }
        return url
    }


    private fun openUrl(url: String?) {
        url?.also {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(it)))
        }
    }

    private fun acceptPrivacyPolicy() {
        logger.info("Policy accepted. Redirecting to DetailedMainView...")
        mListener?.onAcceptPrivacyPolicy()
    }

    private fun rejectPrivacyPolicy() {
        logger.info("Policy rejected. Redirecting to LoginActivity...")
        mListener?.onRejectPrivacyPolicy()
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        fun onAcceptPrivacyPolicy()
        fun onRejectPrivacyPolicy()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(PrivacyPolicyFragment::class.java)

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment PrivacyPolicyFragment.
         */
        fun newInstance(context: Context, state: AppAuthState): PrivacyPolicyFragment {
            return PrivacyPolicyFragment().apply {
                arguments = Bundle().apply {
                    putString(PROJECT_ID_KEY, state.projectId)
                    putString(USER_ID_KEY, state.userId)
                    putString(BASE_URL_PROPERTY, state.getAttribute(BASE_URL_PROPERTY))
                    putString(PRIVACY_POLICY_URL_PROPERTY, state.getAttribute(PRIVACY_POLICY_URL_PROPERTY))
                    putString(PRIVACY_POLICY, context.radarConfig.latestConfig.optString(PRIVACY_POLICY))
                }
            }
        }
    }
}
